### PR TITLE
feat: compiling vision transformers

### LINKED ISSUE #54: NNlib activations cannot be compiled without manual dispatches
```julia
julia> xr = Reactant.ConcreteRArray(rand(Float32, 2, 3))
2×3 Reactant.ConcreteRArray{Float32, (2, 3), 2}:
 0.184252  0.863562  0.0996157
 0.14061   0.574859  0.236953

julia> Reactant.@code_hlo broadcast(tanh, xr)
ERROR: TypeError: in typeassert, expected Tuple{Reactant.TracedRArray{Float32, (2, 3), 2}}, got a value of type Reactant.TracedRArray{Float32, (2, 3), 2}
Stacktrace:
  [1] (::Reactant.var"#398#408"{typeof(broadcast), Reactant.MLIR.IR.Block, Vector{Reactant.TracedRArray}, Tuple{typeof(tanh), Reactant.TracedRArray{Float32, (2, 3), 2}}})()
    @ Reactant /mnt/research/lux/XLA/Reactant.jl/src/utils.jl:105
  [2] block!(f::Reactant.var"#398#408"{typeof(broadcast), Reactant.MLIR.IR.Block, Vector{Reactant.TracedRArray}, Tuple{typeof(tanh), Reactant.TracedRArray{Float32, (2, 3), 2}}}, blk::Reactant.MLIR.IR.Block)
    @ Reactant.MLIR.IR /mnt/research/lux/XLA/Reactant.jl/src/mlir/IR/Block.jl:201
  [3] make_mlir_fn(f::Function, args::Vector{Any}, kwargs::Tuple{}, name::String, concretein::Bool; toscalar::Bool)
    @ Reactant /mnt/research/lux/XLA/Reactant.jl/src/utils.jl:67
  [4] make_mlir_fn
    @ /mnt/research/lux/XLA/Reactant.jl/src/utils.jl:16 [inlined]
  [5] #127
    @ /mnt/research/lux/XLA/Reactant.jl/src/Reactant.jl:1186 [inlined]
  [6] block!(f::Reactant.var"#127#132"{typeof(broadcast), Vector{Any}}, blk::Reactant.MLIR.IR.Block)
    @ Reactant.MLIR.IR /mnt/research/lux/XLA/Reactant.jl/src/mlir/IR/Block.jl:201
  [7] #126
    @ /mnt/research/lux/XLA/Reactant.jl/src/Reactant.jl:1185 [inlined]
  [8] mmodule!(f::Reactant.var"#126#131"{Reactant.MLIR.IR.Module, typeof(broadcast), Vector{Any}}, blk::Reactant.MLIR.IR.Module)
    @ Reactant.MLIR.IR /mnt/research/lux/XLA/Reactant.jl/src/mlir/IR/Module.jl:93
  [9] compile_to_module(mod::Reactant.MLIR.IR.Module, f::Function, args::Vector{Any}; optimize::Bool)
    @ Reactant /mnt/research/lux/XLA/Reactant.jl/src/Reactant.jl:1182
 [10] (::var"#73#74")()
    @ Main /mnt/research/lux/XLA/Reactant.jl/src/Reactant.jl:1314
 [11] context!(f::var"#73#74", ctx::Reactant.MLIR.IR.Context)
    @ Reactant.MLIR.IR /mnt/research/lux/XLA/Reactant.jl/src/mlir/IR/Context.jl:71
 [12] top-level scope
    @ /mnt/research/lux/XLA/Reactant.jl/src/Reactant.jl:1312
 [13] top-level scope
    @ none:1
```

### LINKED ISSUE #135: scalar broadcasting not functional
```julia
using Reactant

x = rand(2, 3)
x_ra = Reactant.to_rarray(x)

function bcast(x)
    sc = sum(x)
    return sc .+ sc
end

bcast_compiled = @compile bcast(x_ra)
```

```julia
f = identity
args = (TracedRArray{Float64,2N}(((:args, 1),)),)
OutShape = (2, 3)
f = +
args = (TracedRArray{Float64,0N}(()), TracedRArray{Float64,0N}(()))
OutShape = ()
ERROR: MethodError: no method matching DenseArrayAttribute(::Vector{Union{}})

Closest candidates are:
  DenseArrayAttribute(::AbstractArray{Float64}; context)
   @ Reactant /mnt/software/lux/Reactant.jl/src/mlir/IR/Attribute.jl:663
  DenseArrayAttribute(::AbstractArray{Float32}; context)
   @ Reactant /mnt/software/lux/Reactant.jl/src/mlir/IR/Attribute.jl:660
  DenseArrayAttribute(::AbstractArray{Int64}; context)
   @ Reactant /mnt/software/lux/Reactant.jl/src/mlir/IR/Attribute.jl:657
  ...

Stacktrace:
  [1] elem_apply(::Function, ::Reactant.TracedRArray{Float64, 0}, ::Reactant.TracedRArray{Float64, 0})
    @ Reactant /mnt/software/lux/Reactant.jl/src/TracedRArray.jl:468
  [2] _copyto!(dest::Reactant.TracedRArray{Float64, 0}, bc::Base.Broadcast.Broadcasted{Nothing, Nothing, typeof(+), Tuple{…}})
    @ Reactant /mnt/software/lux/Reactant.jl/src/TracedRArray.jl:859
  [3] copyto!
    @ /mnt/software/lux/Reactant.jl/src/TracedRArray.jl:765 [inlined]
  [4] copyto!
    @ ./broadcast.jl:967 [inlined]
  [5] copy(bc::Base.Broadcast.Broadcasted{Reactant.AbstractReactantArrayStyle{…}, Nothing, typeof(+), Tuple{…}})
    @ Reactant /mnt/software/lux/Reactant.jl/src/TracedRArray.jl:735
  [6] materialize
    @ ./broadcast.jl:903 [inlined]
  [7] bcast
    @ /mnt/software/lux/Reactant.jl/wip/scalar_bcast.jl:6 [inlined]
  [8] (::Tuple{})(none::Reactant.TracedRArray{Float64, 2})
    @ Base.Experimental ./<missing>:0
  [9] (::Reactant.var"#26#35"{typeof(bcast), Reactant.MLIR.IR.Block, Vector{…}, Tuple{…}})()
    @ Reactant /mnt/software/lux/Reactant.jl/src/utils.jl:100
 [10] block!(f::Reactant.var"#26#35"{typeof(bcast), Reactant.MLIR.IR.Block, Vector{…}, Tuple{…}}, blk::Reactant.MLIR.IR.Block)
    @ Reactant.MLIR.IR /mnt/software/lux/Reactant.jl/src/mlir/IR/Block.jl:201
 [11] make_mlir_fn(f::Function, args::Tuple{…}, kwargs::Tuple{}, name::String, concretein::Bool; toscalar::Bool)
    @ Reactant /mnt/software/lux/Reactant.jl/src/utils.jl:74
 [12] make_mlir_fn
    @ /mnt/software/lux/Reactant.jl/src/utils.jl:24 [inlined]
 [13] #6
    @ /mnt/software/lux/Reactant.jl/src/Compiler.jl:253 [inlined]
 [14] block!(f::Reactant.Compiler.var"#6#11"{typeof(bcast), Tuple{ConcreteRArray{Float64, 2}}}, blk::Reactant.MLIR.IR.Block)
    @ Reactant.MLIR.IR /mnt/software/lux/Reactant.jl/src/mlir/IR/Block.jl:201
 [15] #5
    @ /mnt/software/lux/Reactant.jl/src/Compiler.jl:252 [inlined]
 [16] mmodule!(f::Reactant.Compiler.var"#5#10"{…}, blk::Reactant.MLIR.IR.Module)
    @ Reactant.MLIR.IR /mnt/software/lux/Reactant.jl/src/mlir/IR/Module.jl:93
 [17] compile_mlir!(mod::Reactant.MLIR.IR.Module, f::Function, args::Tuple{ConcreteRArray{Float64, 2}}; optimize::Bool)
    @ Reactant.Compiler /mnt/software/lux/Reactant.jl/src/Compiler.jl:249
 [18] compile_mlir!
    @ /mnt/software/lux/Reactant.jl/src/Compiler.jl:248 [inlined]
 [19] (::Reactant.Compiler.var"#30#32"{typeof(bcast), Tuple{ConcreteRArray{Float64, 2}}})()
    @ Reactant.Compiler /mnt/software/lux/Reactant.jl/src/Compiler.jl:576
 [20] context!(f::Reactant.Compiler.var"#30#32"{typeof(bcast), Tuple{ConcreteRArray{…}}}, ctx::Reactant.MLIR.IR.Context)
    @ Reactant.MLIR.IR /mnt/software/lux/Reactant.jl/src/mlir/IR/Context.jl:71
 [21] compile_xla(f::Function, args::Tuple{ConcreteRArray{Float64, 2}}; client::Nothing)
    @ Reactant.Compiler /mnt/software/lux/Reactant.jl/src/Compiler.jl:573
 [22] compile_xla
    @ /mnt/software/lux/Reactant.jl/src/Compiler.jl:567 [inlined]
 [23] compile(f::Function, args::Tuple{ConcreteRArray{Float64, 2}}; client::Nothing)
    @ Reactant.Compiler /mnt/software/lux/Reactant.jl/src/Compiler.jl:600
 [24] compile(f::Function, args::Tuple{ConcreteRArray{Float64, 2}})
    @ Reactant.Compiler /mnt/software/lux/Reactant.jl/src/Compiler.jl:599
 [25] top-level scope
    @ /mnt/software/lux/Reactant.jl/src/Compiler.jl:360
Some type information was truncated. Use `show(err)` to see complete types.
```

### PR DESCRIPTION
**Pull request by avik-pal**
## Additions / Fixes

- `synchronize` on the result
- support for `batched_mul`
- `getindex` with partial integer indices
- fixes #135 
- fixes #54 

## Update

compiling (mostly) works :tada: 

CUDA does fail atm:

```julia
2024-09-30 15:04:16.435329: I external/xla/xla/stream_executor/cuda/cuda_asm_compiler.cc:399] ptxas warning : Registers are spilled to local memory in function 'gemm_fusion_dot_311', 16 bytes spill stores, 16 bytes spill loads

2024-09-30 15:04:17.975723: I external/xla/xla/stream_executor/cuda/cuda_asm_compiler.cc:399] ptxas warning : Registers are spilled to local memory in function 'gemm_fusion_dot_311_0', 200 bytes spill stores, 192 bytes spill loads

terminate called after throwing an instance of 'xla::XlaRuntimeError'
  what():  INTERNAL: libdevice not found at ./libdevice.10.bc
```

### CODE REVIEW

File: src/TracedRArray.jl [hunk 1]
```diff
@@ -62,14 +62,17 @@ function Base.getindex(
...
+    !isempty(ddims) && return dropdims(x, dims=Tuple(ddims))
+    return x
 end
 
 function Base.view(
     a::TracedRArray{T,N}, indices::Vararg{Union{Base.AbstractUnitRange,Colon},N}
 ) where {T,N}
     # TODO: Implement before merging the PR
-    return error("view is not supported yet")
+    return getindex(a, indices...)
```
Comment (wsmoses) [inline src/TracedRArray.jl:75]: so this isn't quite right since a view won't be mutable

**Comment by avik-pal** [comment]
Yes. I did that to see what other ops need coverage before doing the adapt
stuff we discussed

File: ext/ReactantNNlibExt.jl [hunk 2]
```diff
@@ -206,4 +206,22 @@ end
...
+    if (size(x, 3) != size(y, 3) && size(x, 3) != 1 && size(y, 3) != 1) ||
+        (size(x, 2) != size(y, 1))
+        throw(
+            DimensionMismatch(
+                lazy"size(x) = $(size(x)), size(y) = $(size(y)) inconsistent for batched_matmul.",
+            ),
+        )
+    end
+    if size(x, 3) == size(y, 3)
+        return cat([x[:, :, i] * y[:, :, i] for i in axes(x, 3)]...; dims=Val(3))
```
Comment (wsmoses) [inline ext/ReactantNNlibExt.jl:219]: We should just use dot general here. It supposes batching

File: benchmark/vit.jl [hunk 3]
```diff
@@ -0,0 +1,28 @@
+using Boltz, Lux, Random, LuxCUDA
```
Comment (avik-pal) [inline benchmark/vit.jl:1]: (note to self) convert to a benchmark with ASV for now. later we make it into a proper benchmark on buildkite

File: src/TracedRArray.jl [hunk 4]
```diff
@@ -522,50 +541,26 @@ for (jlop, hloop, hlocomp, merge) in (
...
-                    $merge(elems)
-                end
-            end
-        end
-    else
-        @eval begin
-            function $jlop(
-                @nospecialize(lhs::TracedRArray{T,0}), @nospecialize(rhs::TracedRArray{T,0})
-            ) where {T}
-                return elem_apply($jlop, lhs, rhs)
```
Comment (avik-pal) [inline src/TracedRArray.jl:568]: Base doesn't define these, so deleting

File: ext/ReactantNNlibExt.jl (hunk 2 above)
Comment (avik-pal) [reply ext/ReactantNNlibExt.jl]: done

**Comment by avik-pal** [comment]
benchmark CI won't succeed. The patches needed to run are in this PR

**Review by wsmoses — APPROVED** [review]

**Pull request merged**